### PR TITLE
feat (api): event system

### PR DESCRIPTION
**Pull request by yesnault**

### CODE REVIEW

File: engine/api/event/event.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package event
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/ovh/cds/engine/api/cache"
```
Comment (sguiheux) [inline engine/api/event/event.go]: wrong order

File: engine/api/event/event.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/ovh/cds/sdk"
+
+	"github.com/docker/docker/pkg/namesgenerator"
+	"github.com/spf13/viper"
+)
+
+var hostname, cdsname string
+
+// Routine initializes and run event routine dequeue
+func Routine() {
```
Comment (fsamin) [inline engine/api/event/event.go]: Is it possible to have a better name for this func ?

File: engine/api/event/event.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	cdsname = namesgenerator.GetRandomName(0)
+
+	withKafka := kafkaRoutine()
+
+	for {
+		e := sdk.Event{}
+		cache.Dequeue("events", &e)
+		// send to kafka queue if configured
+		if withKafka {
+			if err := sendOnKafkaTopic(producer, viper.GetString("event_kafka_topic"), &e, log.Debug); err != nil {
```
Comment (fsamin) [inline engine/api/event/event.go]: I prefer not using viper strings in subpackage. we should handle them is main package

File: engine/api/event/kafka.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+var producer sarama.SyncProducer
+
+// Close close event system
+func Close() {
+	closeKafka()
+}
+
+func kafkaRoutine() bool {
+	if viper.GetString("event_kafka_broker_addresses") == "" ||
```
Comment (fsamin) [inline engine/api/event/kafka.go]: I prefer not using viper strings in subpackage. we should handle them is main package

File: engine/api/event/event.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// Routine initializes and run event routine dequeue
+func Routine() {
+	var err error
+	hostname, err = os.Hostname()
+	if err != nil {
+		hostname = fmt.Sprintf("Error while getting Hostname: %s", err.Error())
+	}
+	cdsname = namesgenerator.GetRandomName(0)
+
+	withKafka := kafkaRoutine()
```
Comment (fsamin) [inline engine/api/event/event.go]: How will we deal with other broker system (rabbitmq, redis, websockets and so on...) ?

File: engine/api/event/kafka.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	config.Net.SASL.Password = password
+	config.ClientID = user
+	config.Producer.Return.Successes = true
+
+	producer, errp := sarama.NewSyncProducer(strings.Split(brokerAddresses, ","), config)
+	if errp != nil {
+		return nil, fmt.Errorf("initKafka> Error with init sarama:%s (newSyncProducer on %s user:%s)", errp.Error(), brokerAddresses, user)
+	}
+
+	InfoLogFunc("initKafka> Kafka used at %s on topic:%s", brokerAddresses, topic)
```
Comment (fsamin) [inline engine/api/event/kafka.go]: InfoLogFunc can be nil

File: engine/api/event/kafka.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	return producer, nil
+}
+
+// sendOnKafkaTopic send a hook on a topic kafka
+func sendOnKafkaTopic(producer sarama.SyncProducer, topic string, event *sdk.Event, DebugLogFunc func(string, ...interface{})) error {
```
Comment (fsamin) [inline engine/api/event/kafka.go]: InfoLogFunc and DebugLogFunc can be nil

File: engine/api/event/publish.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+// PublishActionBuild sends a actionBuild event
+func PublishActionBuild(pb *sdk.PipelineBuild, ab *sdk.ActionBuild) {
+	e := sdk.EventJob{
+		Version:         pb.Version,
+		JobName:         ab.ActionName,
+		Status:          ab.Status,
+		Queued:          ab.Queued.Unix(),
+		Start:           ab.Start.Unix(),
+		Done:            ab.Done.Unix(),
+		Model:           ab.Model,
```
Comment (fsamin) [inline engine/api/event/publish.go]: is it model ID or model Name ?

File: engine/api/event/publish.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+		PipelineName:    pb.Pipeline.Name,
+		PipelineType:    pb.Pipeline.Type,
+		ProjectKey:      pb.Pipeline.ProjectKey,
+		ApplicationName: pb.Application.Name,
+		EnvironmentName: pb.Environment.Name,
+		BranchName:      pb.Trigger.VCSChangesBranch,
+		Hash:            pb.Trigger.VCSChangesHash,
+	}
```
Comment (fsamin) [inline engine/api/event/publish.go]: Maybe repository manager and repo fullname are usefull

File: engine/api/event/publish.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+		Done:        pb.Done.Unix(),
+		RepositoryManagerName: rmn,
+		PipelineName:          pb.Pipeline.Name,
+		PipelineType:          pb.Pipeline.Type,
+		ProjectKey:            pb.Pipeline.ProjectKey,
+		ApplicationName:       pb.Application.Name,
+		EnvironmentName:       pb.Environment.Name,
+		BranchName:            pb.Trigger.VCSChangesBranch,
+		Hash:                  pb.Trigger.VCSChangesHash,
+	}
```
Comment (fsamin) [inline engine/api/event/publish.go]: Maybe repository manager and repo fullname are usefull

File: engine/api/repositoriesmanager/events.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+			}
+		}
+	}
+}
+
+func processEvent(db database.Querier, event sdk.Event) error {
+	log.Debug("repositoriesmanager>processEvent> receive: type:%s all: %+v", event.EventType, event)
+
+	var eventpb sdk.EventPipelineBuild
+	if event.EventType == fmt.Sprintf("%T", sdk.EventPipelineBuild{}) {
```
Comment (fsamin) [inline engine/api/repositoriesmanager/events.go]: early return plz

File: engine/api/repositoriesmanager/events.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+		if err := mapstructure.Decode(event.Payload, &eventpb); err != nil {
+			log.Warning("Error during consumption: %s", err)
+			return nil
+		}
+	} else {
+		// skip all event != eventPipelineBuild
+		return nil
+	}
+
+	if eventpb.RepositoryManagerName == "" {
```
Comment (fsamin) [inline engine/api/repositoriesmanager/events.go]: this statement should be perform at first

**Comment by fsamin** [comment]
Can we have a cli action go stream events ?

**Comment by fsamin** [comment]
I did not see any tests :(

**Comment by fsamin** [comment]
what about /mon/status ?

File: engine/api/event/publish.go (hunk 8 above)
Comment (yesnault) [inline engine/api/event/publish.go]: Same as ActionBuild.Model -> renamed in Event, next commit.

File: engine/api/event/event.go (hunk 2 above)
Comment (yesnault) [inline engine/api/event/event.go]: fixed in next commit.

File: engine/api/event/event.go (hunk 5 above)
Comment (yesnault) [inline engine/api/event/event.go]: fixed in next commit.

File: engine/api/event/kafka.go (hunk 4 above)
Comment (yesnault) [inline engine/api/event/kafka.go]: fixed in next commit.

File: engine/api/event/kafka.go (hunk 6 above)
Comment (yesnault) [inline engine/api/event/kafka.go]: fixed in next commit.

File: engine/api/event/kafka.go (hunk 7 above)
Comment (yesnault) [inline engine/api/event/kafka.go]: fixed in next commit.

File: engine/api/repositoriesmanager/events.go (hunk 11 above)
Comment (yesnault) [inline engine/api/repositoriesmanager/events.go]: fixed in next commit

File: engine/api/repositoriesmanager/events.go (hunk 12 above)
Comment (yesnault) [inline engine/api/repositoriesmanager/events.go]: eventpb != event, eventpb not available before

File: engine/api/event/publish.go (hunk 10 above)
Comment (yesnault) [inline engine/api/event/publish.go]: RepositoryManagerName already here, fullname added in next commit.

File: engine/api/event/publish.go (hunk 9 above)
Comment (yesnault) [inline engine/api/event/publish.go]: actionBuildNotif -> no repo manager, only for EventPipelineBuild.

**Pull request merged**